src/cheri_cap_common.sail

```diff
@@ -88,41 +228,48 @@ function capToMemBits(cap) : Capability -> CapBits =
 function memBitsToCapability(tag, b) : (bool, CapBits) -> Capability =
   capBitsToCapability(tag, b ^ null_cap_bits)
 
-
-
-function getCapBoundsBits(c) : Capability -> (CapAddrBits, CapLenBits) =
+function getCapBoundsBitsUncorrected(c) : Capability -> (CapLenBits, CapLenBits) =
   let E = min(maxE, unsigned(c.E)) in
   let a : CapAddrBits = c.address in
+  let a_ext : CapLenBits = EXTZ(a) in
+  let a_top = truncateLSB(a_ext, cap_len_width - E - mantissa_width) in
+  let base : CapLenBits = a_top @ c.B @ zeros(E) in
+  let top  : CapLenBits = a_top @ c.T @ zeros(E) in
+    (base, top)
+
+function getCapLegalised(c) : Capability -> Capability =
+  let (base, top) = getCapBoundsBitsUncorrected(c) in
+  /* Check that base is within the address space */
+  if unsigned(c.E) <= maxE
+   & (base <_u (to_bits(cap_len_width, 1) << cap_addr_width)) then
+    c
+  else
+    /* Legalise the capability by zeroing the ie bit, implicitly setting E to 0. TODO avoid 90 magic number */
```

Comment:
I don't see a 90 magic number?

Reply:
Maybe that's why it's magic :P but I assume this used to do a `1 << 90` to set IE in the compressed 128-bit representation.

Reply:
Yep, good spot. I'll remove the TODO